### PR TITLE
Fix the failure to lint modules contained under an identically named directory

### LINKED ISSUE #4444: Linting fails if module contains module of the same name
### Steps to reproduce

Given multiple files:
```
.
`-- a/
    |-- a.py
    `-- b.py
```
Which are all empty, running `pylint a` fails:

```
$ pylint a
************* Module a
a/__init__.py:1:0: F0010: error while code parsing: Unable to load file a/__init__.py:
[Errno 2] No such file or directory: 'a/__init__.py' (parse-error)
$
```

However, if I rename `a.py`, `pylint a` succeeds:

```
$ mv a/a.py a/c.py
$ pylint a
$
```
Alternatively, I can also `touch a/__init__.py`, but that shouldn't be necessary anymore.

### Current behavior

Running `pylint a` if `a/a.py` is present fails while searching for an `__init__.py` file.

### Expected behavior

Running `pylint a` if `a/a.py` is present should succeed.

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0a3
astroid 2.5.6
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0]
```

### Additional info

This also has some side-effects in module resolution. For example, if I create another file `r.py`:

```
.
|-- a
|   |-- a.py
|   `-- b.py
`-- r.py
```

With the content:

```
from a import b
```

Running `pylint -E r` will run fine, but `pylint -E r a` will fail. Not just for module a, but for module r as well.

```
************* Module r
r.py:1:0: E0611: No name 'b' in module 'a' (no-name-in-module)
************* Module a
a/__init__.py:1:0: F0010: error while code parsing: Unable to load file a/__init__.py:
[Errno 2] No such file or directory: 'a/__init__.py' (parse-error)
```

Again, if I rename `a.py` to `c.py`, `pylint -E r a` will work perfectly.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
Fixes #4444: modules contained under an identically named directory weren't linted (failed with `parse-error`).

## Notes
This PR causes a test failure in `TestImportsChecker.test_relative_beyond_top_level_two`, but that test uses `epylint` to excute pylint (!) which does additional manipulation of `cwd`, so I don't think it's correct.

Notice that linting the package in the test fails using regular pylint, but the failure is now more accurate on this branch:

**main**
```shell
$ pylint tests/regrtest_data/beyond_top_two
************* Module beyond_top_two
tests/regrtest_data/beyond_top_two/__init__.py:1:0: F0010: error while code parsing: Unable to load file tests/regrtest_data/beyond_top_two/__init__.py:
[Errno 2] No such file or directory: 'tests/regrtest_data/beyond_top_two/__init__.py' (parse-error)
```
**PR**
```shell
$ pylint tests/regrtest_data/beyond_top_two
************* Module tests/regrtest_data/beyond_top_two
tests/regrtest_data/beyond_top_two:1:0: F0001: No module named tests/regrtest_data/beyond_top_two (fatal)
```

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
- [ ] Test this against #3651

File: pylint/lint/expand_modules.py [hunk 1]
```diff
@@ -103,9 +105,7 @@ def expand_modules(
                 )
                 if filepath is None:
                     continue
-            except (ImportError, SyntaxError) as ex:
-                # The SyntaxError is a Python bug and should be
-                # removed once we move away from imp.find_module: https://bugs.python.org/issue10588
```
Comment (Pierre-Sassoulas) [inline pylint/lint/expand_modules.py:108]: 👍

File: tests/checkers/unittest_imports.py [hunk 2]
```diff
@@ -40,6 +41,9 @@ def test_relative_beyond_top_level(self) -> None:
             self.checker.visit_importfrom(module.body[2].body[0])
 
     @staticmethod
+    @pytest.mark.xfail(
+        reason="epylint manipulates cwd; these tests should not be using epylint"
```
Comment (Pierre-Sassoulas) [inline tests/checkers/unittest_imports.py:45]: Shouldn't we fix it or remove it ? What's the benefit of keeping xfailed tests ?

File: tests/lint/unittest_lint.py [hunk 3]
```diff
@@ -942,3 +942,12 @@ def test_lint_namespace_package_under_dir(initialized_linter: PyLinter) -> None:
         create_files(["outer/namespace/__init__.py", "outer/namespace/module.py"])
         linter.check(["outer.namespace"])
     assert not linter.stats.by_msg
+
+
+def test_identically_named_nested_module(initialized_linter: PyLinter) -> None:
+    with tempdir():
+        create_files(["identical/identical.py"])
+        with open("identical/identical.py", "w", encoding="utf-8") as f:
+            f.write("import imp")
```
Comment (Pierre-Sassoulas) [inline tests/lint/unittest_lint.py:951]: I wonder where the idea for this example came from 😄

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty good already, I have a question regarding the xfailed test.

File: tests/checkers/unittest_imports.py (hunk 2 above)
Comment (jacobtylerwalls) [reply tests/checkers/unittest_imports.py:45]: Yeah, I just didn't want to block this PR on resolving that. All of the tests in that file probably shouldn't be using epylint. Would you like me to open an issue?

Comment (Pierre-Sassoulas) [reply tests/checkers/unittest_imports.py:45]: Yeah let's open an issue, I'm not sure emacs is used a lot so it's probably not worth a fix, no one read the emacs documentation at least.

Comment (jacobtylerwalls) [reply tests/checkers/unittest_imports.py:45]: #7252

Comment (Pierre-Sassoulas) [reply tests/checkers/unittest_imports.py:45]: Thank you !

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**